### PR TITLE
[feat] experimental MEVO layer

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?

This introduces a custom NN layer called MemoryEfficientVocabOutput (MEVO). This layer includes a linear output projection and a crossentropy loss (softmax + NLL loss). The main goal for this custom layer is to support very large scale training with large batch size, long sequence length and large vocabulary size.

The techniques uses in this layer are the following:

1. custom fused, checkpointed, tiled functions that avoiding materializing a HUGE output matrix from the output projection.
2. custom backward to avoid multiple copies of gradient buffers.
3. custom FSDP wrapping to shard the parameter and gradients for the projection.

A benchmark and 2 unit tests are added to test this layer. The layer is compared with a baseline implementation for both correctness, memory and speed. The speed is largely reduced due to complex python pytorch operator sequences. In the future, we will address the speed issues. At this moment, due to reduced memory, we can increase the batch size for the overall model training and gain moderate speedup overall.

The associated FSDP change is already merged with #836. It might be useful to refer to that commit while reviewing.

Anyone is welcome to review and provide constructive feedback. Thanks!

This is experimental and once it is more validated and more ready for general consumption, I will add more extensive user doc.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
This is ready for review. Please take a look whoever has the time. No pressure.

It is still experimental since I have yet to run end to end tests with loss curves. Unit tests do ensure numerical comparisons though. Also, speed is not great. More work is needed in the future, likely with Triton.

I will have a separate doc with detailed performance and memory analysis.

File: fairscale/experimental/nn/ssd_offload.py [hunk 1]
```diff
@@ -166,9 +166,9 @@ def copy_into_tensor(self, tensor: torch.Tensor) -> None:
         """Copies SsdTensorHandle's data into the given tensor.
 
         If the tensor is in memory, this function copies the data
-        into the passed in tensor. Otherwise, it reads from file into tensor, 
+        into the passed in tensor. Otherwise, it reads from file into tensor,
```
Comment (min-xu-ai) [inline fairscale/experimental/nn/ssd_offload.py:169]: these whilespace change is added by pre-commit hook.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1412,7 +1412,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
             if hasattr(param._linked_param, "_is_shared") and param._linked_param._is_shared:
                 param = param._linked_param
 
-        assert param.grad is not None
+        assert param.grad is not None, param.shape
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1415]: this single line is for better debugging output.

File: fairscale/experimental/nn/mevo.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+        """Forward function that computes the max, without saving activations."""
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            print("DEBUG max fwd")
+        ctx.save_for_backward(i, w)
+        ctx.kernel_obj = kernel_obj
+        ctx.w_idx = w_idx
+        ctx.w_split_size = w_split_size
+        ctx.args = {}
+        assert split_dim == 0
+        with torch.no_grad():
```
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:168]: this is to save some memory/speed, because autograd is wrong there ? curious as of why this is needed basically

File: fairscale/experimental/nn/mevo.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+            print("DEBUG max bwd")
+        assert len(args) == 1
+        # Gradients should already exist due to TargetScoreFunction's backward.
+        assert ctx.kernel_obj.proj_weight.grad is not None
+
+        # Get saved i and w.
+        i, w = ctx.saved_tensors
+        assert i.requires_grad
+        assert w.requires_grad
+        i = i.detach().requires_grad_(True)
```
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:187]: the sequence here (detach, FW/BW with autograd, then no_grad again) is pretty confusing without much context. Could you add some explanations ? I'm a bit worried about future maintainability

File: fairscale/experimental/nn/mevo.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+        # nlprob, then sum over all tokens.
+        return -prob.sum()
+
+    def forward(self, *input: Any, **kwargs: Any) -> Any:
+        if DEBUG and dist.is_initialized() and dist.get_rank() == 0:
+            cur_mem = round(torch.cuda.memory_allocated() / 1024 / 1024)
+            mem = round(torch.cuda.max_memory_allocated() / 1024 / 1024)
+            print("DEBUG cur, peak", cur_mem, mem)
+        assert kwargs == {}
+        input, target = input
```
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:430]: I don't get that, why not declare them directly as inputs and get some help from mypy ?

Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:430]: thanks for the review, Ben!

if I recall this correctly, the prototype for `forward` in mypy is actually the Any, Any one. That's why I used that to make mypy happier. If you know a way that I can use separate args more directly here without too much mypy magic, that'd be great!

File: fairscale/experimental/nn/mevo.py (hunk 4 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:187]: Yes, very true! I will add more comments in the code around this.

File: fairscale/experimental/nn/mevo.py (hunk 3 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:168]: This is for saving memory with checkpointing. I will have an extensive doc that describes the algorithm in more details. Will add more comment here too.

File: fairscale/experimental/nn/mevo.py (hunk 5 above)
Comment (blefaudeux) [reply fairscale/experimental/nn/mevo.py:430]: oh I typically ignore the mypy warning on this line ( #type: ignore), because I think that declaring inputs helps with readability (a lot of people will look at the interface, a lot more than the number of people who will look at the code in any case). The issue with the inputs goes with Pytorch not being typed in general I think, so this #type: ignore is a little ugly but I can live with it

**Comment by min-xu-ai** [comment]
@blefaudeux, addressed your comments. Thanks for the review!

If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.

**Comment by anj-s** [comment]
> @blefaudeux, addressed your comments. Thanks for the review!
> 
> If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.



> @blefaudeux, addressed your comments. Thanks for the review!
> 
> If others no major issues, I will merge pretty soon. That will unblock my work of integrating it with fairseq properly.

Taking a look now, would be great if you could wait till tomorrow morning. thanks!

**Comment by min-xu-ai** [comment]
> Taking a look now, would be great if you could wait till tomorrow morning. thanks!

Thanks for taking a look! Of course!

File: fairscale/experimental/nn/mevo.py [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+    """ Baseline softmax that does an output linear projection and a softmax.
+
+        This is intended to be used with an embedding layer with shared weights.
+
+    Args:
+        proj_weight (nn.Parameter):
+            The shared weight.
+        tile_factor (int):
+            Unused. It is here to make kernel init easier with MEVO.
+        log (bool):
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:78]: nit: I would say "use_log" instead of "log". makes it more clear that it is a bool.

File: fairscale/experimental/nn/mevo.py [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
+        """Forward that directly compute the loss."""
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        x = super().forward(input, target)
+        x = F.nll_loss(x, target, reduction="sum")
+        return x
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:130]: nit: early return in line 129?

File: fairscale/experimental/nn/mevo.py [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+
+
+class GetMaxFunction(torch.autograd.Function):
+    """Custom checkpointed function to get max-per-token from an input and a weight"""
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:134]: nit: missing period at the end of this sentence.

File: fairscale/experimental/nn/mevo.py [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+        return max_n
+    return n
+
+
+def _reshape_inputs(input: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Convert 3D inputs to 2D for this kernel"""
+    if len(input.shape) == 3:
+        input = input.reshape(-1, input.shape[2])
+    if len(target.shape) == 2:
+        target = target.reshape(-1)
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:46]: will this be a 1-D or 2D target?

File: fairscale/experimental/nn/mevo.py [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+
+
+class GetMaxFunction(torch.autograd.Function):
+    """Custom checkpointed function to get max-per-token from an input and a weight"""
+
+    @staticmethod
+    def get_max(i: torch.Tensor, w: torch.Tensor, fp: bool) -> torch.Tensor:
+        """ i: (split-of-tokens, d_model), w: (split-of-vocabs, d_model)"""
+        _m = torch.matmul(i, w.T)
+        if fp:
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:140]: what does fp stand for?

File: fairscale/experimental/nn/mevo.py (hunk 4 above)
Comment (anj-s) [reply fairscale/experimental/nn/mevo.py:190]: thanks for adding the comments! they definitely help but I think we need a doc to be honest :) this is pretty complex and context is pretty low with just the code.

File: fairscale/experimental/nn/mevo.py [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+            grads = torch.split(ctx.kernel_obj.proj_weight.grad, ctx.w_split_size)
+            grads[ctx.w_idx].add_(w.grad)
+        return i.grad, None, maxs.grad, None, None, None, None
+
+
+class TargetScoreFunction(torch.autograd.Function):
+    """Custom checkpointed function to compute the target score."""
+
+    @staticmethod
+    def get_target_score(i: torch.Tensor, w: torch.Tensor, target: torch.Tensor, fp: bool) -> torch.Tensor:
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:280]: can we rename i and w across this file to something that it represents? it might be easier to grok this code.

File: fairscale/experimental/nn/mevo.py [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+        self.trigger._linked_param = linked_param
+
+    def forward(self) -> torch.Tensor:  # type: ignore
+        return BackwardTriggerFn.apply(self.trigger._linked_param, self.trigger)
+
+
+class MemoryEfficientVocabOutput(nn.Module):  # AKA. MEVO
+    """ Fused fc + softmax + nll_loss in a tiled fashion.
+
+        This uses much less memory but is quite a bit slower.
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:388]: just curious about why it is slower. do we know why?

File: fairscale/experimental/nn/mevo.py [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+    """ Baseline that does an output projection, a softmax & a NLL loss (cross-entropy).
+
+        See BaselineSoftmax above. Constructor is the same. Only difference is in the
+        forward function.
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:123]: Why do we implement this? is it for testing convenience?

File: fairscale/experimental/nn/mevo.py [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+        tile_factor (int):
+            Number of splits to use on the input sequence and vocab dimensions.
+        reduction (str):
+            Reduction OP (sum or mean).
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 16, reduction: str = "sum"):
+        super().__init__()
+        self.proj_weight = proj_weight
+        self.tf_in, self.tf_w = tile_factor, tile_factor
```
Comment (anj-s) [inline fairscale/experimental/nn/mevo.py:402]: does tile_factor have to be the same? how does it impact memory savings?

File: tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        # LNs are not strictly needed for this test, but they help reduce the loss quickly
+        # and improves the numerical stability.
+        self.ln1 = nn.LayerNorm(D_MODEL).cuda().half()
+        self.ln2 = nn.LayerNorm(D_MODEL).cuda().half()
+
+        if with_fsdp:
+            # Shared layers much be un-flatten.
+            self.l0 = FSDP(self.l0, flatten_parameters=False, mixed_precision=False, compute_dtype=torch.float16)
+            self.l1 = FSDP(self.l1, flatten_parameters=False, mixed_precision=False, compute_dtype=torch.float16)
+            self.l1.append_shared_param(self.l0.module.weight)
```
Comment (anj-s) [inline tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py:56]: MEVO should also work for non-shared layers right? Curious what the reason was for having that support.

**Comment by anj-s** [comment]
nice PR @min-xu-ai ! Would love to see a doc when you get a chance. Feel free to merge since my comments are more questions with a few nits. Without a doc I am not sure I can comment about the algorithm anyway. Thanks for waiting for the merge.

**Comment by min-xu-ai** [comment]
> nice PR @min-xu-ai ! Would love to see a doc when you get a chance. Feel free to merge since my comments are more questions with a few nits. Without a doc I am not sure I can comment about the algorithm anyway. Thanks for waiting for the merge.

Thank you so much for the review @anj-s! Yes, the doc is coming too. The comments and suggestions are great. Will address them!

File: tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py (hunk 15 above)
Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_shared_weights_mevo.py:56]: The floor memory consumption won't be as good otherwise. If mevo is used without this sharding support, then its param and grad memory will stay unsharded (i.e. full) for the entire forward/backward pass. This reduces the memory usable by other layers. With this support, we shard them after the layer is done computing, leaving much more memory to other layers, like the attention layers.

File: fairscale/experimental/nn/mevo.py (hunk 14 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:402]: It doesn't need to be the same. I haven't extensively tuning it. Will add a TODO.

File: fairscale/experimental/nn/mevo.py (hunk 13 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:123]: correct. This is purely for testing and benchmarking.

Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:123]: added a comment.

File: fairscale/experimental/nn/mevo.py (hunk 12 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:388]: Yes, we know some of it is due to re-computation. Others are likely due to python looping. This is definitely an important future work.

File: fairscale/experimental/nn/mevo.py (hunk 10 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:140]: fp is full precision, sharing the same acronym with floating point. :-)

I will make those longer names.

File: fairscale/experimental/nn/mevo.py (hunk 9 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:46]: target is always 1D in this layer. We just reshape the input when the input is 2D. input is also always reshaped from 3d to 2d.

File: fairscale/experimental/nn/mevo.py (hunk 7 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:130]: sure.

File: fairscale/experimental/nn/mevo.py (hunk 11 above)
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:280]: Alright, I added an explanation of them at the first occurrence.

File: fairscale/experimental/nn/mevo.py [hunk 16]
```diff
@@ -165,6 +161,9 @@ def forward(  # type: ignore
         ctx.w_split_size = w_split_size
         ctx.args = {}
         assert split_dim == 0
+        # During forward, we use ``no_grad'' to avoid saving the activations.
```
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:170]: <3

File: fairscale/experimental/nn/mevo.py [hunk 17]
```diff
@@ -184,15 +183,21 @@ def backward(ctx: Any, *args: Any) -> Any:
         i, w = ctx.saved_tensors
         assert i.requires_grad
         assert w.requires_grad
+        # We use ``detach()'' to ensure the backward call below does not
```
Comment (blefaudeux) [inline fairscale/experimental/nn/mevo.py:192]: it really helps, I never realized that custom backward functions could themselves build a new autograd graph, I always assumed (I was wrong) that autograd was hands off there.. thanks for the comments @min-xu-ai !

**Review by blefaudeux — APPROVED** [review]
Looks great to me, feels like some of that could be sped up either with the recent cuda graph optimization from pytorch or by lowering this into triton or cuda kernels, but step by step is a good approach I think !

**Comment by min-xu-ai** [comment]
> Looks great to me, feels like some of that could be sped up either with the recent cuda graph optimization from pytorch or by lowering this into triton or cuda kernels, but step by step is a good approach I think !

Couldn’t agree more!
